### PR TITLE
feat: delete user by email

### PR DESCRIPTION
**Pull request by ZainabAlayande**
## How this should be manually tested
- Ensure the user is authenticated and authorized to delete an account.
- Access the route as an authenticated user `(ROLE_SUPER_ADMIN)`.
- Send a `DELETE` request to `/api/v1/users/delete` with the following request body to delete a user by their email:

#### Request
```
{
    "email": "user.email@example.com"
}
```

#### Responses
```
{
    "status_code": "success",
    "message": "The account has been successfully deleted."
}
```

```
{
    "status_code": "404",
    "message": "User not found with email: user.email@example.com"
}
```

```
{
    "status_code": "400",
    "message": "Bad Request. The email field is required."
}
```

## Checklist of What I Did:

- [ ] Implemented the `DELETE /api/v1/users/delete` endpoint to delete a user by email.
- [ ] Added validation to ensure the email field is provided and is not empty.
- [ ] Checked if the user exists by their email and handled cases where the user does not exist.
- [ ] Implemented error handling to return appropriate status codes and messages.
- [ ] Wrote unit tests to validate the service logic for various scenarios (successful deletion, user not found, bad request).

## Reference Issue
[https://github.com/hngprojects/hng_boilerplate_java_web/issues/417]()

### CODE REVIEW

**Review by lurldgbodex — APPROVED** [review]
can't find your test

**Review by lurldgbodex — CHANGES_REQUESTED** [review]
Did you write test?

File: src/main/java/hng_java_boilerplate/user/controller/UserController.java [hunk 1]
```diff
@@ -30,12 +30,24 @@ public ResponseEntity<?> getUserDetails(@PathVariable String userId) {
     @PreAuthorize("hasRole('ROLE_SUPER_ADMIN')")
     @GetMapping(value = "/members", produces = "application/json")
     public ResponseEntity<?> getAllMembers(@RequestParam int page, Authentication authentication) {
+        if (authentication == null) return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
```
Comment (Adewale66) [inline src/main/java/hng_java_boilerplate/user/controller/UserController.java:34]: As long as the route is protected, the authentication object will never be null which also means it's not needed

File: src/main/java/hng_java_boilerplate/user/controller/UserController.java [hunk 2]
```diff
@@ -30,12 +30,24 @@ public ResponseEntity<?> getUserDetails(@PathVariable String userId) {
...
+    @PreAuthorize("hasRole('ROLE_SUPER_ADMIN')")
+    @DeleteMapping("/delete")
+    public ResponseEntity<?> deleteUserByEmail(@RequestBody DeleteUserRequest request, Authentication authentication) {
+        if (authentication == null) throw new UnAuthorizedException("Unauthorized");
+        Response<?> response = userService.deleteUserByEmail(request, authentication);
+        return ResponseEntity.status(HttpStatus.OK).body(response);
+    }
+
+
     @GetMapping("/me/{id}")
```
Comment (Adewale66) [inline src/main/java/hng_java_boilerplate/user/controller/UserController.java:51]: What is this endpoint supposed to return?

File: src/main/java/hng_java_boilerplate/user/serviceImpl/UserServiceImpl.java [hunk 3]
```diff
@@ -382,6 +381,22 @@ public List<MembersResponse> getAllUsers(int page, Authentication authentication
         return users;
     }
 
+
+    @Transactional
+    public Response<?> deleteUserByEmail(DeleteUserRequest request, Authentication authentication) {
+        String email = request.getEmail();
+        if (email == null || email.isEmpty()) {
```
Comment (Adewale66) [inline src/main/java/hng_java_boilerplate/user/serviceImpl/UserServiceImpl.java]: You can add a NotBlank annotation on your DeleteUserRequest object

**Review by Adewale66 — CHANGES_REQUESTED** [review]

File: src/main/java/hng_java_boilerplate/user/controller/UserController.java (hunk 2 above)
Comment (ZainabAlayande) [reply src/main/java/hng_java_boilerplate/user/controller/UserController.java:51]: Retrieve a single user object

Comment (Adewale66) [reply src/main/java/hng_java_boilerplate/user/controller/UserController.java:51]: if it's just retrieving a single user then the route should be /{id}

**Review by Adewale66 — CHANGES_REQUESTED** [review]

**Review by Adewale66 — APPROVED** [review]
LGTM

**Pull request merged**